File: ios/NoSpam/DeviceIdentity.swift
import Foundation
import CryptoKit

/// Per-install random identifier, sent only as an irreversible hash — never
/// IDFA/device serials — so reports can be rate-limited without tracking users.
enum DeviceIdentity {
    private static let storageKey = "nospam.device_identifier"

    static var hashedDeviceId: String {
        let digest = SHA256.hash(data: Data(rawDeviceId().utf8))
        return digest.map { String(format: "%02x", $0) }.joined()
    }

    private static func rawDeviceId() -> String {
        if let existing = UserDefaults.standard.string(forKey: storageKey) {
            return existing
        }
        let newId = UUID().uuidString
        UserDefaults.standard.set(newId, forKey: storageKey)
        return newId
    }
}

File: android/app/src/main/java/com/konrad/nospam/DeviceIdentity.kt
package com.konrad.nospam

import android.content.Context
import java.security.MessageDigest
import java.util.UUID

/**
 * Per-install random identifier, sent only as an irreversible hash — never
 * the Android ID or ad ID — so reports can be rate-limited without tracking users.
 */
object DeviceIdentity {
    private const val PREFS_NAME = "nospam_prefs"
    private const val KEY_DEVICE_ID = "device_identifier"

    fun hashedDeviceId(context: Context): String {
        val digest = MessageDigest.getInstance("SHA-256")
            .digest(rawDeviceId(context).toByteArray(Charsets.UTF_8))
        return digest.joinToString("") { "%02x".format(it) }
    }

    private fun rawDeviceId(context: Context): String {
        val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
        prefs.getString(KEY_DEVICE_ID, null)?.let { return it }
        val newId = UUID.randomUUID().toString()
        prefs.edit().putString(KEY_DEVICE_ID, newId).apply()
        return newId
    }
}
